⚠️ Serve catalog over HTTPS

trgeiger opened the pull request:
<!--
Please prefix the title of this PR with one of the following icons:

    * ⚠ (:warning:, major/breaking change)
    * ✨ (:sparkles:, minor/compatible change)
    * 🐛 (:bug:, patch/bug fix)
    * 📖 (:book:, docs)
    * 🌱 (:seedling:, other)

-->
This adds the ability to use HTTPS on the catalog server.

If the TLS cert and key filenames are provided via the relevant flags, then the server will be set to use HTTPS. If neither or only one of those options are provided, the server will use HTTP.

This also changes the http-external-address flag to just external-address, with the user no longer prepending the address with http or https.
 - Should I add a check for if the user *does* have http or https in the address, and adjust it to the proper one depending on the presence of the key/cert?
 - Should we add a duplicate flag --http-external-address for backwards compatibility and just map it to the new flag if used?

This PR also includes an overhaul of the organization of the manifests. Following the example of rukpak, cert-manager is used in the Makefile build targets so HTTPS is enabled in kind, but cert-manager is not a required dependency of catalogd. Instead, it's applied as an overlay which we target in the Makefile.

everettraven committed:
make catalog server serve catalog contents over HTTPS

adds cert-manager as a dependency again to create self-signed
certs for the catalog server

Signed-off-by: everettraven <everettraven@gmail.com>

everettraven committed:
fix e2e

Signed-off-by: everettraven <everettraven@gmail.com>

trgeiger committed:
Reorganize manifests for cert-manager overlay

This allows the use of alternate certificate managers.

Signed-off-by: Tayler Geiger <tayler@redhat.com>

trgeiger committed:
Reconfigure TLS functionality to use Listener

Fix a few manifest issues as well.

Signed-off-by: Tayler Geiger <tayler@redhat.com>

everettraven commented on cmd/manager/main.go:
```diff
@@ -143,14 +150,40 @@ func main() {
 
-	baseStorageURL, err := url.Parse(fmt.Sprintf("%s/catalogs/", httpExternalAddr))
+	if certFile != "" && keyFile != "" {
+		cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+		if err != nil {
+			setupLog.Error(err, "unable to load certificate key pair")
+			os.Exit(1)
+		}
+		config := &tls.Config{
+			Certificates: []tls.Certificate{cert},
+			MinVersion:   tls.VersionTLS13,
+		}
```
I think @joelanford raised a concern in my previous draft PR about handling certificate rotations. Cert-manager should auto-rotate the certificates but I think we will need to implement specific logic to be able to pick up the auto-rotated certs (the file on disk should be updated since the secret created by cert-manager is mounted as a volume). 

I think we can use the `tls.Config.GetCertificate` field here to dynamically load and return the certificate on every request.

Bonus points would be to create a struct that watches for file changes and caches the certificate so we aren't having to do the certificate loading every time, but I think that could be an optimization made in the future and doesn't _have_ to be done in this PR.

trgeiger replied:
@everettraven I've taken a shot at implementing the file watcher in latest push if you want to take a look

joelanford commented on cmd/manager/main.go:
```diff
@@ -143,14 +150,40 @@ func main() {
 	}
 
-	baseStorageURL, err := url.Parse(fmt.Sprintf("%s/catalogs/", httpExternalAddr))
+	if certFile != "" && keyFile != "" {
```
I think if only one is set, we should error out instead of silently ignoring the one flag that was set.

joelanford commented on internal/third_party/server/server.go:
```diff
@@ -38,7 +38,7 @@ var (
+// Server is a general purpose HTTP(S) server Runnable for a manager.
 // It is used to serve some internal handlers for health probes and profiling,
 // but it can also be used to run custom servers.
 type Server struct {
```
I'd like to eventually remove this code. Could you add an issue for bumping controller-runtime to 0.18.x, which has an exported server runnable?

trgeiger replied:
#265 created

trgeiger committed:
Add certwatcher for TLS cert and key from controller-runtime

- Add error for missing either tls-key or tls-cert arguments.
- Move server creation and configuration to serverutil

Signed-off-by: Tayler Geiger <tayler@redhat.com>

everettraven commented on cmd/manager/main.go:
```diff
@@ -143,14 +151,48 @@ func main() {
+			os.Exit(1)
+		}
+		if err := mgr.Add(&tlsFileWatcher); err != nil {
+			setupLog.Error(err, "unable to start TLS file watcher")
```
nit: I don't think the call to `mgr.Add` would actually start the TLS file watcher and would instead add it to a slice of `Runnables` that are started in a later call to `mgr.Start()`

trgeiger replied:
Ah so I should just adjust the error message here, or should I not be adding it to the manager?

everettraven replied:
Still need to add it to the manager, would suggest adjusting the error message here to prevent confusion between "adding" and "starting" the watcher

everettraven commented on config/base/rbac/role.yaml:
```diff
@@ -0,0 +1,65 @@
+- apiGroups:
+  - ""
+  resources:
+  - pods
```
Out of scope for this PR, but seeing this I don't think we need these permissions on pods anymore. we don't use a pod to unpack the catalog contents anymore and probably forgot to come back and cleanup these permissions. I'll create an issue for this.

trgeiger commented on internal/certwatcher/tls_cert_watcher.go:
```diff
@@ -0,0 +1,87 @@
+		case <-ctx.Done():
+			return ctx.Err()
+		case event := <-watcher.Events:
+			if event.Has(fsnotify.Write) {
```
Good catch and good question. I'll do some digging/testing

joelanford commented on cmd/manager/main.go:
```diff
@@ -143,14 +151,48 @@ func main() {
+		}
+		externalAddr = "https://" + externalAddr
+	default:
+		setupLog.Error(nil, "unable to configure TLS certificates, both tls-cert and tls-key arguments are required")
```
Nit: "are required" makes it sound like both flags are always required. Maybe soften it a bit to try to clarify that these are both or neither.

```suggestion
		setupLog.Error(nil, "unable to configure TLS certificates, --tls-cert and --tls-key flags are required together")
```

joelanford commented on cmd/manager/main.go:
```diff
@@ -143,14 +151,48 @@ func main() {
 	}
 
-	baseStorageURL, err := url.Parse(fmt.Sprintf("%s/catalogs/", httpExternalAddr))
+	switch {
```
Maybe put this in a separate function that returns `*net.Listener, extrenalAddr, error`?

joelanford commented on cmd/manager/main.go:
```diff
@@ -162,10 +204,11 @@ func main() {
 
 	if err := mgr.Add(&catalogServer); err != nil {
-		setupLog.Error(err, "unable to start catalog server")
+		setupLog.Error(err, "problem adding catalog server to manager")
```
Nit: consistent error messages.

```suggestion
		setupLog.Error(err, "unable to add catalog server to manager")
```

joelanford commented on internal/certwatcher/tls_cert_watcher.go:
```diff
@@ -0,0 +1,87 @@
+
+var _ manager.Runnable = (*TLSCertificateWatcher)(nil)
+
+type TLSCertificateWatcher struct {
```
It looks like controller-runtime already has a certificate watcher that we can use: https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/certwatcher

(Sorry: I didn't remember this existed until I googled for it)

trgeiger replied:
oh wow I even named it the same lol

trgeiger replied:
Okay I've implemented certwatcher from controller-runtime. I put the new helper function for creating the listener into a new internal package that replaces my tls-file-watcher stuff. Let me know if I should put that somewhere else instead. I didn't want to put it into the third-party/server stuff since we want to remove that eventually.

joelanford commented on internal/serverutil/serverutil.go:
```diff
@@ -0,0 +1,42 @@
+		}
+		config := &tls.Config{
+			GetCertificate: tlsFileWatcher.GetCertificate,
+			MinVersion:     tls.VersionTLS13,
```
Is it reasonable to expect all clients to be able to use TLS 1.3 at this point?

everettraven replied:
Looks like TLS 1.2 is the minimum by default (see [source](https://cs.opensource.google/go/go/+/refs/tags/go1.22.3:src/crypto/tls/common.go;drc=059a9eedf45f4909db6a24242c106be15fb27193;l=738)). I'd feel better about this if we are just sticking with the default that the std lib is setting here as they likely have more knowledge in what is a reasonable expectation a larger set of clients may have

joelanford commented on internal/serverutil/serverutil.go:
```diff
@@ -0,0 +1,42 @@
+		addr = "https://" + addr
+		return listener, addr, nil
+	default:
+		return nil, "", fmt.Errorf("unable to configure TLS certificates, tls-cert and tls-key flags must be used together")
```
As nice as it is for this to be the `default` case, it means that we're doing flag validation fairly late in the process. WDYT about having a specific check in `main()` for the XOR cases that we do right after `flag.Parse()`?

everettraven replied:
+1 to doing flag validation as early in the process as possible

joelanford commented on internal/serverutil/serverutil.go:
```diff
@@ -0,0 +1,42 @@
+	"sigs.k8s.io/controller-runtime/pkg/certwatcher"
+)
+
+func ConfigureListener(cert string, key string, addr string, catalogAddr string, mgr ctrl.Manager) (net.Listener, string, error) {
```
As I mentioned in Slack, I didn't realize how complex the signature for this function would be when I suggested it. At this point, it seems like we might as well go one step further and setup the entire server as part of this function.

I think just adding the `storeDir` would give us all the info we need to do that?

Also some nit suggestions on naming and order of parameters.
- Put the manager parameter first. That aligns with other functions that configure/add stuff to the manager.
- Group `storeDir` and `externalAddr` together, since they are used to build the storage struct.
- Use `certFile` and `keyFile` (instead of `cert` and `key` to make it more clear that they are the filenames of the cert and key, not the cert and key themselves)
- Group listenAddr, certFile, and keyFile together since they are used to build the listener

Putting all that together, WDYT of this?
```
func AddCatalogServerToManager(mgr ctrl.Manager, storeDir, externalAddr, listenAddr, certFile, keyFile string) error
```

everettraven replied:
If we are concerned about the size of the function signature, it can probably be reduced to something like:
```go
type CatalogServerConfig struct {
...
}

func AddCatalogServerToManager(mgr ctrl.Manager, cfg CatalogServerConfig) (net.Listener, string, error) { 
... 
}
```

trgeiger commented on cmd/manager/main.go:
```diff
@@ -103,6 +106,18 @@ func main() {
+		os.Exit(1)
+	}
+
+	protocol := "http://"
```
Since localStorage is used outside of just the server setup, I felt like I had to put this logic here. It's not *too* redundant but happy to implement a more elegant suggestion for handling the creation of the localStorage. It's needed for the catalogServer as well as the catalogReconciler.

everettraven replied:
This seems fine to me

everettraven commented on internal/serverutil/serverutil.go:
```diff
@@ -0,0 +1,69 @@
+func AddCatalogServerToManager(mgr ctrl.Manager, cfg CatalogServerConfig) error {
+	listener, err := net.Listen("tcp", cfg.CatalogAddr)
+	if err != nil {
+		return err
```
Could we wrap all the errors returned by this function with some more context? For example, for this error I'd like to see something along the lines of:
```suggestion
		return fmt.Errorf("creating listener for %q: %w", cfg.CatalogAddr, err)
```

This will help with troubleshooting any problems since we would be able to more easily tell from the logs where exactly in this function the error was returned

trgeiger replied:
done, adjusted the top-level error in main.go as well

everettraven reviewed (dismissed):
LGTM - thanks @trgeiger . I'll hold merging for a bit to allow @joelanford to take another pass

trgeiger commented:
Thanks, Bryce. Should we adjust anything in the readme's quickstart instructions or anywhere else in docs?

everettraven commented:
> Thanks, Bryce. Should we adjust anything in the readme's quickstart instructions or anywhere else in docs?

Ah, great question (and good catch). We should probably update the quickstart instructions for port-forwarding and the `curl` command for fetching catalog contents to reflect the new state with an HTTPS based server as opposed to HTTP. I think this information is also present in https://github.com/operator-framework/catalogd/blob/main/docs/fetching-catalog-contents.md

everettraven commented:
>How should we handle changing the instructions, given that HTTP is still supported if no TLS keys are provided, and the cert-manager is now an overlay. Do we just add that information as caveats to the instructions? Should I detail pulling down the cert and providing it to curl? etc.

I'm not too familiar with how the overlays work, but IIUC the default deployment (i.e what is released) is going to be including the cert-manager resources and enable TLS (please correct me if I am wrong). If that _is_ the default I would document whatever the default is.

I would not document pulling down the cert and providing it to curl and instead document with `curl -k` to ignore the certs. We can include a disclaimer in the docs that this is only for demonstration purposes and that if they want to trust the certificates to follow some other documentation on how to do that (I'm sure there is some documentation on this that already exists)

trgeiger committed:
Update README and docs for HTTPS

trgeiger commented:
I made a couple small edits to README.md and https://github.com/operator-framework/catalogd/blob/main/docs/fetching-catalog-contents.md, let me know what you think

everettraven commented on docs/fetching-catalog-contents.md:
```diff
@@ -1,12 +1,16 @@
 # Fetching `Catalog` contents from the Catalogd HTTP Server
 This document covers how to fetch the contents for a `Catalog` from the
-Catalogd HTTP Server that runs when the `HTTPServer` feature-gate is enabled
+Catalogd HTTP(S) Server that runs when the `HTTPServer` feature-gate is enabled
```
While you are making changes here, I now see that this hasn't been updated in some time. There is no longer a `HTTPServer` feature gate as it is the default behavior (and has been for some time). I'll leave it up to you if you want to fix it in this PR or create an issue to track bringing this doc up to date.

everettraven commented on docs/fetching-catalog-contents.md:
```diff
@@ -1,12 +1,16 @@
 - A `Catalog` named `operatorhubio` has been created and successfully unpacked
 (denoted in the `Catalog.Status`)
 
+**NOTE:** By default, Catalogd is configured to use TLS with self-signed certificates for demonstration purposes.
```
```suggestion
**NOTE:** By default, Catalogd is configured to use TLS with self-signed certificates.
```

everettraven reviewed (dismissed):
A couple nits on the docs edits but I won't hold this PR on those. Great work @trgeiger !

trgeiger commented:
Those were quick fixes, just made em so no need to hold off for another issue/pr for those.
